packages/gatsby-source-contentful/src/source-nodes.js

```diff
@@ -319,6 +295,107 @@ export async function sourceNodes(
+          foreignReference => {
+            const { name, id: contentfulId, type, spaceId } = foreignReference
+
+            const nodeId = createNodeId(
+              makeId({
+                spaceId,
+                id: contentfulId,
+                type,
+                currentLocale: n.node_locale,
+                defaultLocale,
+              })
+            )
+
+            // Create new reference field when none exists
+            if (!n[name]) {
+              existingNodesThatNeedReverseLinksUpdateInDatastore.add(n)
+              n[name] = [nodeId]
+              return
+            }
+
+            // Add non existing references to reference field
+            if (n[name] && !n[name].includes(nodeId)) {
+              existingNodesThatNeedReverseLinksUpdateInDatastore.add(n)
+              n[name].push(nodeId)
+            }
+          }
+        )
+      }
+
+      // Remove references to deleted nodes
+      if (n.contentful_id && deletedEntryGatsbyReferenceIds.size) {
+        Object.keys(n).forEach(name => {
+          // @todo Detect reference fields based on schema. Should be easier to achieve in the upcoming version.
+          if (!name.endsWith(`___NODE`)) {
+            return
+          }
+          if (Array.isArray(n[name])) {
+            n[name] = n[name].filter(referenceId => {
+              const shouldRemove =
+                deletedEntryGatsbyReferenceIds.has(referenceId)
+              if (shouldRemove) {
+                existingNodesThatNeedReverseLinksUpdateInDatastore.add(n)
+              }
+              return !shouldRemove
+            })
+          } else {
+            const referenceId = n[name]
+            if (deletedEntryGatsbyReferenceIds.has(referenceId)) {
+              existingNodesThatNeedReverseLinksUpdateInDatastore.add(n)
+              n[name] = null
+            }
+          }
+        })
+      }
+    })
+
+  // We need to call `createNode` on nodes we modified reverse links on,
+  // otherwise changes won't actually persist
+  if (existingNodesThatNeedReverseLinksUpdateInDatastore.size) {
+    for (const node of existingNodesThatNeedReverseLinksUpdateInDatastore) {
+      function addChildrenToList(node, nodeList = [node]) {
+        for (const childNodeId of node?.children ?? []) {
+          const childNode = getNode(childNodeId)
+          if (childNode) {
+            // important
+            nodeList.push(childNode)
+            addChildrenToList(childNode)
+          }
+        }
+        return nodeList
+      }
+
+      const nodeAndDescendants = addChildrenToList(node)
+      for (const nodeToUpdate of nodeAndDescendants) {
+        // we need to remove properties from existing fields
+        // that are reserved and managed by Gatsby.
+        delete nodeToUpdate.internal.owner
+
+        // TODO: we need to recalculate `internal.contentDigest`
+        // for now just modyfing it to ensure gatsby picks up data change
+        node.internal.contentDigest += `X`
```

Comment:
We need to change `contentDigest`, but not sure to what. Now we use `.sys.updateAt`, but this won't change (it was other node linking to this one that changed). I can add counter or something instead. 

Here how we set this up in full node creation mode:
https://github.com/gatsbyjs/gatsby/blob/65e9e31b216a8405d4c9b8dd7c32fd454effa1d5/packages/gatsby-source-contentful/src/normalize.js#L678-L679